executor/src/witgen/jit/affine_symbolic_expression.rs

```diff
@@ -257,41 +254,55 @@ impl<T: FieldElement, V: Ord + Clone + Display> AffineSymbolicExpression<T, V> {
 
         // Check if they are mutually exclusive and compute assignments.
         let mut covered_bits: <T as FieldElement>::Integer = 0.into();
-        let mut effects = vec![];
-        for (var, coeff, constraint) in constrained_coefficients {
-            let mask = *constraint.multiple(coeff).mask();
-            if !(mask & covered_bits).is_zero() {
+        let mut components = vec![];
+        for (variable, coeff, constraint) in constrained_coefficients {
+            let is_negative = !coeff.is_in_lower_half();
+            let coeff_abs = if is_negative { -coeff } else { coeff };
+            let Some(exponent) = log2_exact(coeff_abs.to_arbitrary_integer()) else {
+                // We could work with non-powers of two, but it would require
+                // division instead of shifts.
+                return Ok(ProcessResult::empty());
+            };
+            let bit_mask = *constraint.multiple(coeff_abs).mask();
+            if !(bit_mask & covered_bits).is_zero() {
                 // Overlapping range constraints.
                 return Ok(ProcessResult::empty());
             } else {
-                covered_bits |= mask;
+                covered_bits |= bit_mask;
             }
-            let masked = -&self.offset & mask;
-            effects.push(Effect::Assignment(
-                var.clone(),
-                masked.integer_div(&coeff.into()),
-            ));
+            components.push(BitDecompositionComponent {
+                variable,
+                // We negate here because we are solving
+                // c_1 * x_1 + c_2 * x_2 + ... + offset = 0,
+                // instead of
+                // c_1 * x_1 + c_2 * x_2 + ... = offset.
+                is_negative: !is_negative,
+                exponent: exponent as u64,
+                bit_mask,
+            });
         }
 
         if covered_bits >= T::modulus() {
             return Ok(ProcessResult::empty());
         }
 
-        // We need to assert that the masks cover "-offset",
-        // otherwise the equation is not solvable.
-        // We assert -offset & !masks == 0
-        if let Some(offset) = self.offset.try_to_number() {
-            if (-offset).to_integer() & !covered_bits != 0.into() {
-                return Err(Error::ConflictingRangeConstraints);
+        if !components.iter().any(|c| c.is_negative) {
+            // If all coefficients are positive and the offset is known, we can check
```

Comment:
I mean if the offset is a known number, we can actually determine that in all cases, but the ones with negative coefficients are more complicated...